File: core/osrd-railjson/src/main/java/fr/sncf/osrd/railjson/schema/common/RJSWaypointRef.java
package fr.sncf.osrd.railjson.schema.common;

import com.squareup.moshi.Json;
import java.util.Objects;

@ExcludeFromGeneratedCodeCoverage
public final class RJSWaypointRef<T extends Identified> {
    public ID<T> id;

    public enum RJSWaypointType {
        @Json(name = "BufferStop")
        BUFFER_STOP,
        @Json(name = "Detector")
        DETECTOR
    }

    public RJSWaypointType type;

    public RJSWaypointRef(ID<T> id, RJSWaypointType type) {
        this.id = id;
        this.type = type;
    }

    public RJSWaypointRef(String id, RJSWaypointType type) {
        this(new ID<>(id), type);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, type);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null)
            return false;
        if (obj.getClass() != RJSWaypointRef.class)
            return false;
        var o = (RJSWaypointRef<?>) obj;
        return id.equals(o.id) && type.equals(o.type);
    }

    @Override
    public String toString() {
        return String.format("RJSWaypointRef { type=%s, id=%s }", type, id.id);
    }
}

File: core/osrd-railjson/src/main/java/fr/sncf/osrd/railjson/schema/common/RJSWaypointRef.kt
package fr.sncf.osrd.railjson.schema.common

import com.squareup.moshi.Json
import fr.sncf.osrd.railjson.schema.infra.trackobjects.RJSRouteWaypoint
import java.util.*

@ExcludeFromGeneratedCodeCoverage
data class RJSWaypointRef<T : RJSRouteWaypoint>(var id: ID<T>, var type: RJSWaypointType) {
    enum class RJSWaypointType {
        @Json(name = "BufferStop") BUFFER_STOP,
        @Json(name = "Detector") DETECTOR,
    }

    constructor(id: String, type: RJSWaypointType) : this(ID<T>(id), type)
}
